File: uniconn-js/core/src/p2p/protocol.ts
/**
 * P2P Control Messages for Relay Signaling
 */

export const MsgType = {
  ANNOUNCE: 'ANNOUNCE',
  FIND_PEER: 'FIND_PEER',
  FOUND_PEER: 'FOUND_PEER',
  RELAY_REQ: 'RELAY_REQ',
  INCOMING_RELAY: 'INCOMING_RELAY',
  ACCEPT_RELAY: 'ACCEPT_RELAY',
  RELAY_ACK: 'RELAY_ACK',
  ERROR: 'ERROR'
} as const;

export type MsgTypeString = typeof MsgType[keyof typeof MsgType];

export interface Envelope<T = any> {
  type: MsgTypeString;
  payload: T;
}

export interface AnnouncePayload {
  fingerprint: string;
  direct_addresses: string[];
}

export interface FindPeerPayload {
  target_fingerprint: string;
}

export interface FoundPeerPayload {
  target_fingerprint: string;
  direct_addresses: string[];
}

export interface RelayReqPayload {
  target_fingerprint: string;
}

export interface RelayAckPayload {
  target_fingerprint: string;
  success: boolean;
  reason?: string;
}

export interface IncomingRelayPayload {
  requester_fingerprint: string;
  session_token: string;
}

export interface AcceptRelayPayload {
  session_token: string;
}

File: uniconn-py/uniconn/p2p/protocol.py
from typing import Any, Dict, List, TypedDict

class MsgType:
    ANNOUNCE = "ANNOUNCE"
    FIND_PEER = "FIND_PEER"
    FOUND_PEER = "FOUND_PEER"
    RELAY_REQ = "RELAY_REQ"
    INCOMING_RELAY = "INCOMING_RELAY"
    ACCEPT_RELAY = "ACCEPT_RELAY"
    RELAY_ACK = "RELAY_ACK"
    ERROR = "ERROR"

class Envelope(TypedDict):
    type: str
    payload: Any

class AnnouncePayload(TypedDict):
    fingerprint: str
    direct_addresses: List[str]

class FindPeerPayload(TypedDict):
    target_fingerprint: str

class FoundPeerPayload(TypedDict):
    target_fingerprint: str
    direct_addresses: List[str]

class RelayReqPayload(TypedDict):
    target_fingerprint: str

class RelayAckPayload(TypedDict, total=False):
    target_fingerprint: str
    success: bool
    reason: str

class IncomingRelayPayload(TypedDict):
    requester_fingerprint: str
    session_token: str

class AcceptRelayPayload(TypedDict):
    session_token: str
